sotodlib/toast/ops/intensity_templates.py

```diff
@@ -0,0 +1,454 @@
+# a BSD-style license that can be found in the LICENSE file.
+
+import os
+import pickle
+import re
+import warnings
+from time import time
+
+import h5py
+import numpy as np
+import traitlets
+from astropy import units as u
+
+from toast import qarray as qa
+from toast.mpi import MPI, Comm, MPI_Comm
+from toast.observation import default_values as defaults
+from toast.timing import function_timer
+from toast.traits import Int, Unicode, trait_docs
+from toast.utils import Environment, Logger
+from toast.ops import Operator
+from toast.ops.demodulation import Lowpass
+
+
+XAXIS, YAXIS, ZAXIS = np.eye(3)
+
+
+@trait_docs
+class IntensityTemplates(Operator):
+    """Operator that derives intensity templates to use in destriping or
+    filtering.  Typically run after demodulation or pair differencing
+    when detector data is either pure intensity or pure polarization.
+
+    """
+
+    API = Int(0, help="Internal interface version for this operator")
+
+    det_data = Unicode(
+        defaults.det_data, help="Observation detdata key to use as input"
+    )
+
+    det_mask = Int(
+        defaults.det_mask_nonscience,
+        help="Bit mask value for per-detector flagging",
+    )
+
+    det_flags = Unicode(
+        defaults.det_flags,
+        allow_none=True,
+        help="Observation detdata key for flags to use",
+    )
+
+    det_flag_mask = Int(
+        defaults.det_mask_nonscience,
+        help="Bit mask value for detector sample flagging",
+    )
+
+    shared_flags = Unicode(
+        defaults.shared_flags,
+        allow_none=True,
+        help="Observation shared key for telescope flags to use",
+    )
+
+    shared_flag_mask = Int(
+        defaults.shared_mask_nonscience,
+        help="Bit mask value for optional shared flagging",
+    )
+
+    view = Unicode(
+        None,
+        allow_none=True,
+        help="Use this view of the data in all observations",
+    )
+
+    cache_dir = Unicode(
+        None,
+        allow_none=True,
+        help="If specified, cache templates in this directory.",
+    )
+
+    pattern = Unicode(
+        f"^demod0.*",
+        allow_none=True,
+        help="Regex pattern to match against detector names. Only detectors "
+        "that match the pattern are used to derive intensity templates."
+    )
+
+    mode = Unicode(
+        "radius:1deg",
+        help="Method of deriving the intensity template. Supported modes:\n"
+        "radius:<Quantity> -- Average the intensity in the disc.  Use a small "
+        "value to match only one pixel\n"
+        "all -- Average all detectors matching `pattern`"
+        )
+
+    submode = Unicode(
+        "raw",
+        help="Post processing of the intensity template. Supported modes\n"
+        "raw -- no post processing\n"
+        "lowpass:<Quantity> -- low-pass the intensity template\n"
+        "split:<Quantity>,<Quantity>,... -- split the template into several "
+        "complementary templates at given frequencies\n",
+    )
```

Comment:
I know this code is going to be refactored, but for consideration:  rather than passing a complicated set of options as a string to a single trait, it might be more clear and easier to set in the calling code and config files if there were multiple traits.  Something like:
```
method = Unicode("radius", help="Method of deriving the template ('radius' or 'all')")

radius = Quantity(1.0 * u.deg, help="If using method=radius, average the intensity in a disc")

process = Unicode(None, allow_none=True, help="Post processing to apply to templates ('lowpass', 'split')")

process_lowpass = Quantity(10.0 * u.Hz, help="If process='lowpass', the cutoff frequency")

process_split = List([], help="If process='split', a list of Quantities specifying the frequencies of the breaks")
```